### PR TITLE
Support UNIX domain sockets in Windows RS5

### PR DESCRIPTION
**Pull request by damonbarry**
Windows supports UNIX domain sockets, but only on RS5 or later. So this change includes logic in CMakeLists.txt which adds the compiler definition `AF_UNIX_ON_WINDOWS` when building on RS5 or later. It also makes the Winsock2 adapter for socketio look a lot more like the Berkeley adapter, including support for the `ADDRESS_TYPE` option in `socketio_setoption`.

### CODE REVIEW

**Comment by damonbarry** [comment]
@jspaith I moved the parsing logic into socketio_win32 as discussed. Please review...

File: adapters/socketio_win32.c [hunk 1]
```diff
@@ -245,6 +251,94 @@ void socketio_destroy(CONCRETE_IO_HANDLE socket_io)
...
+    {
+#endif
+        char portString[16];
+        ADDRINFO* addr_info = NULL;
+        ADDRINFO addrHint = { 0 };
+        addrHint.ai_family = AF_INET;
+        addrHint.ai_socktype = SOCK_STREAM;
+        addrHint.ai_protocol = 0;
+
+        sprintf(portString, "%u", socket_io_instance->port);
```
Comment (dcristoloveanu) [inline adapters/socketio_win32.c]: >sprintf [](start = 8, length = 7)

This can fail

File: adapters/socketio_win32.c [hunk 2]
```diff
@@ -245,6 +251,94 @@ void socketio_destroy(CONCRETE_IO_HANDLE socket_io)
...
+        if (getaddrinfo(socket_io_instance->hostname, portString, &addrHint, &addr_info) != 0)
+        {
+            LogError("Failure: getaddrinfo failure %d.", WSAGetLastError());
+            result = __FAILURE__;
+        }
+        else
+        {
+            result = connect_socket(socket_io_instance->socket, addr_info->ai_addr, sizeof(*addr_info->ai_addr));
```
Comment (dcristoloveanu) [inline adapters/socketio_win32.c]: >connect_socket [](start = 21, length = 14)

Don't you want to log the error when this fails?

File: adapters/socketio_win32.c [hunk 3]
```diff
@@ -245,6 +251,94 @@ void socketio_destroy(CONCRETE_IO_HANDLE socket_io)
...
+        }
+
+        if (path_len + 1 > sizeof(addr_un.sun_path))
+        {
+            LogError("Path '%s' is too long for a unix socket (max len = %zu)", path, sizeof(addr_un.sun_path));
+            result = __FAILURE__;
+        }
+        else
+        {
+            memset(&addr_un, 0, sizeof(addr_un));
```
Comment (dcristoloveanu) [inline adapters/socketio_win32.c]: >memset [](start = 12, length = 6)

(void)

File: adapters/socketio_win32.c (hunk 2 above)
Comment (damonbarry) [reply adapters/socketio_win32.c]: connect_socket is an internal function, and it logs its own failures.

File: adapters/socketio_win32.c (hunk 3 above)
Comment (damonbarry) [reply adapters/socketio_win32.c]: Will do

File: adapters/socketio_win32.c (hunk 1 above)
Comment (damonbarry) [reply adapters/socketio_win32.c]: I'll add the error handler for this

Comment (damonbarry) [reply adapters/socketio_win32.c]: Fixed

File: adapters/socketio_win32.c (hunk 3 above)
Comment (damonbarry) [reply adapters/socketio_win32.c]: Fixed

**Pull request merged**